src/fields.rs

```diff
@@ -6,3 +6,11 @@ mod fq;
 
 pub use fp::*;
 pub use fq::*;
+
+#[cfg(feature = "gpu")]
+fn u64_to_u32(limbs: &[u64]) -> alloc::vec::Vec<u32> {
+    limbs
+        .iter()
+        .flat_map(|limb| alloc::vec![(limb & 0xFFFF_FFFF) as u32, (limb >> 32) as u32])
```

Comment:
This performs a transient allocation per limb, which is unnecessary. Instead:
```suggestion
        .flat_map(|limb| Some((limb & 0xFFFF_FFFF) as u32).into_iter().chain(Some((limb >> 32) as u32)))
```

Once our MSRV reaches 1.56 and we enable 2021 edition, we can change this to:
```suggestion
        .flat_map(|limb| [(limb & 0xFFFF_FFFF) as u32, (limb >> 32) as u32].into_iter())
```

Reply:
That's really a nice one, thanks!